### DISCUSSION TITLE
Can microsecond precision be kept in firestore_web?

### DISCUSSION THREAD

**Discussion by ditman** [firebase/flutterfire#4731] [Feature request]
As seen in [this bug](https://github.com/flutter/flutter/issues/73334), when converting Firestore Timestamps to a web DateTime, the code is just using milliseconds. See code [here](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/cloud_firestore/cloud_firestore_web/lib/src/interop/utils/utils.dart#L26), and [here](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/cloud_firestore/cloud_firestore_web/lib/src/interop/utils/utils.dart#L39), losing microsecond precision if written by a mobile client.

However, it seems that DateTime may support `fromMicrosecondsSinceEpoch` ([docs](https://api.flutter.dev/flutter/dart-core/DateTime/DateTime.fromMicrosecondsSinceEpoch.html)) and the JS Timestamp seems to store up to `nanosecons` for Timestamps ([code](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/cloud_firestore/cloud_firestore_web/lib/src/interop/firestore_interop.dart#L386)). It also has a constructor that allows Timestamps to be created from seconds + nanoseconds.

**Would it be possible to keep microsecond precision for Timestamps in the the web implementation of Firestore?**

**Comment by Constans** [comment]
I stumbled upon this issue as I'm using the Timestamp to DateTime and the `millisecondsSinceEpoch` as part of the encryption key used to decode some data. The solution/hack is to retry decoding and makeup for the variation of the value on milliseconds (+/-1). 

The issue seems to be happening only on Flutter web. On mobile it works fine.

Not a great solution but with this approach it manages to decode the encrypted data after 3 max retries. Sometimes it works with the 1st run. 

Still an issue after 3+ years...
